torch/csrc/jit/script/sugared_value.cpp

```diff
@@ -67,22 +67,29 @@ std::shared_ptr<SugaredValue> SimpleValue::attr(
           Symbol::aten(builtin_cast_methods().at(field)),
           NamedValue(loc, "self", value_));
     }
-    // functions that are just direct property lookups on tensor
-    // must be registered as prim::<name>(Tensor t) -> <return_type>
-    static const std::unordered_set<std::string> fields = {
-        "dtype",
-        "device",
-        "grad",
-        "data",
-        "shape",
-        "is_cuda",
-        "is_sparse",
-        "is_mkldnn",
-        "is_quantized",
-        "requires_grad",
-        "layout",
-    };
-    if (fields.count(field)) {
+  }
+  // accessing properties of Tensor and Device that are implemented as
+  // prim:: operators
+  static const std::unordered_map<TypeKind, std::unordered_set<std::string>>
+      builtin_properties = {{TypeKind::TensorType,
+                             {
+                                 "dtype",
+                                 "device",
+                                 "grad",
+                                 "data",
+                                 "shape",
+                                 "is_cuda",
+                                 "is_sparse",
+                                 "is_mkldnn",
+                                 "is_quantized",
+                                 "requires_grad",
+                                 "layout",
+                             }},
+                            {TypeKind::DeviceObjType, {"type", "index"}}};
```

Comment:
why is there "index" on device type here ?

Reply:
forgot to implement it, it's added in the latest commit